Generate: add model class validation

gante opened the pull request:
# What does this PR do?

Fixes #18210 

This PR adds model class validation at the start of generate (all model classes inherit `GenerationMixin`, but few can use `generate()`). It also adds an exception that attempts to redirect the users to the right classes.

gante committed:
add model class validation

sgugger reviewed (approved):
Thanks for fixing!

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -841,6 +841,16 @@ def compute_transition_beam_scores(
+        if not hasattr(self, "prepare_inputs_for_generation"):
+            model_class = self.__class__.__name__
+            raise TypeError(
+                f"The current model class ({model_class}) is not compatible with `.generate()`, as it doesn't have a ",
```
(nit) Wondering if it could make sense to make a really nifty error message here where we check if any class of the model type of the given class in in one of the auto classes below - maybe over-engineered though

patrickvonplaten reviewed (approved):
Sounds good! Wondering if it could make sense to make the error message a bit more detailed E.g. if someone uses `BertModel` and gets an error that one should use one of the following AutoModel classes, I'm not sure how helpful this is). Could be super cool that if one uses `BertModel` and gets as an answer that `BertForCausalLM` supports `generate` it could be a nicer UX. But maybe over engineered.

gante committed:
tentative commit (let's see if tests break)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18902). All of your documentation changes will be reflected on that endpoint.

gante committed:
exception points to the right class

gante commented:
@sgugger @patrickvonplaten I've requested a re-review of this PR. As per @patrickvonplaten's suggestion, the PR was upgraded to contain the exact class the user should use in the exception (as opposed to pointing to all generate-compatible auto classes).

In the process of building it, I've noticed that the previous version of this PR was incorrect anyways -- PT and TF had a default `prepare_inputs_for_generation`, so we couldn't rely on its existence. Only 1 model was using this default, so I removed it and implemented it in the missing model. The default `prepare_inputs_for_generation` was a public method, but since this PR blocks the use of `generate()` with classes that are not intended to be used with it anyways, removing the public method should have little impact. Nevertheless, it is a point to consider in the review!

_______________________________

Here's an example with the current version of the PR:
```
>>> from transformers import AutoModel
>>> model = AutoModel.from_pretrained("distilgpt2")
>>> model.generate("foo")
TypeError: The current model class (GPT2Model) is not compatible with `.generate()`, as it doesn't have a language model head. Please use one of the following classes instead: {'GPT2LMHeadModel'}
```

LysandreJik reviewed (approved):
Looks good to me! I haven't tested thoroughly but from a few quick tests it seems to work well.

Thanks @gante!

sgugger reviewed (approved):
Oh missed the notification that re-requested the review. Still LGTM!
